### ISSUE TITLE
lmlsqnonlin prompts error due to undefined values

### ISSUE THREAD

**Issue by ShonaHepworth** [luisfabib/DeerAnalysis2#97]
**I have the following MATLAB script to analyse my 4 pulsed DEER data, changing the file name each time to match that of the saved file:**

````matlab
clear,clc;

[t,V] = deerload('FileName.DSC'); %File name
axis tight, grid on;
ylabel('(V(t) (AU)'),xlabel('tus');

V = correctphase(V);
t = correctzerotime(V,t);
V = correctscale(V,t);

clf,plot(t,V,'k.')
axis tight, grid on
ylabel('(V(t) (AU)'),xlabel('tus');

r = linspace(2,5,150);
fitsignal(V,t,r,'P',@bg_hom3d,@ex_4pdeer);
````

**For some files, I keep getting the following...**

````
Error using eig
Input matrix contains NaN or Inf.

Error in lmlsqnonlin (line 346)
    MaxEigJJ = max(eig(JJ));

Error in fitparamodel (line 371)
    [parfit,fval,~,exitflag]  =
    solverFcn(@ResidualsFcn,StartParameters,lowerBounds,upperBounds,solverOpts);

Error in correctscale (line 87)
parfit = fitparamodel(V_,fitmodel,t_,par0,'Upper',ub,'Lower',lb,'Rescale',false);

Error in untitled2 (line 9)
V = correctscale(V,t);
````

**I am unsure how to fix this or what is going on, the script works with some files but comes up with the above message for others. Thanks**

**Comment by laenan8466** [comment]
Hey,
could you add the DSC/DTA files for failing runs?

**Comment by luisfabib** [comment]
Hi Shona, 

could you please add one example DSC/DTA file which prompts the error, as well as the DeerLab release version you are using? 

The error is arising from one of the toolbox-free solvers which are somewhat undertested by DeerLab's test suit and might have missed a specific case such as yours leading to the error. 

If I can reproduce the error I will be able to fix this quicker. Thank you :)

**Comment by ShonaHepworth** [comment]
Hi,

I was going to answer online via github but it won’t let me attach the DCS or DTA files for my data so hopefully this email will find you okay!

I am using DeerLab 0.9.1 and MATLAB R2017_b.

Thanks,

Shona

**Comment by luisfabib** [comment]
I am afraid I did not get the DTA/DSC files attached in the email. Maybe if you put them in a ZIP-file it will be easier to send. Otherwise you can send them via email to my email address directly.

**Comment by luisfabib** [comment]
Using your data I could not reproduce the error (MATLAB R2019b or R2016b, Win10OS), with these solvers there is often a certain sensitivity to the computer specifics. However, I wrote I small brute-force script to try to trigger the conditions under which the error above was being prompted. 

I have identified and fixed several things which might be causing these errors. Still, since I could not reproduce the problem with your data I am not completely sure of having fixed it. 

 Shona, could you please:
  1) Download this [lmlsqnonlin.zip](https://github.com/JeschkeLab/DeerLab/files/4700855/lmlsqnonlin.zip) and extract the  ``lmlsqnonlin.m`` file
  2) Copy it into the ``DeerLab\functions\private`` folder (replacing the old one)
  3) Run your script with the data that was leading an error before

Please let me know if the error persists or if it has been solved. In case of the latter I will release an official patch with the fix, otherwise I will keep debugging the error.

**Comment by ShonaHepworth** [comment]
I am still getting the same error unfortunately. Could this be a Mac problem? As I asked someone in the same lab to try it on their windows laptop and it worked fine.

**Comment by luisfabib** [comment]
Ok, I have the feeling that the issue here is that in the algorithm of ` lmlsqnonlin.m` there are some values which are dependent on the precision of the computer. Some of these values are then being fed through a sequence of function handles which end up leading to NaN values. These conditions would explain the erratic behavior of the error and the fact that it is computer-dependent. 

I might have traced the error and have made a modification. @ShonaHepworth I must to ask for your help again. Could you please: 
  1) Download this new [lmlsqnonlin.zip](https://github.com/JeschkeLab/DeerLab/files/4702769/lmlsqnonlin.zip) and extract again the  ``lmlsqnonlin.m`` file
  2) Copy it into the ``DeerLab\functions\private`` folder (replacing the old one)
  3) Run your script with the data that was leading to an error before
  4) If the error persists try to replace in your script the line

          fitsignal(V,t,r,'P',@bg_hom3d,@ex_4pdeer);
       
      by the line
         
          fitsignal(V,t,r,@dd_gauss,@bg_hom3d,@ex_4pdeer);

     and please let me know it that leads to the same error again. (If not it will point me out to where the error is actually ocurring)

Thank you very much for you help!!

**Comment by ShonaHepworth** [comment]
If I use the original line of code mentioned ( fitsignal(V,t,r,'P',@bg_hom3d,@ex_4pdeer); )… MATLAB doesn’t come up with the error immediately, and the script just runs for a long time. I have left it for about 30 minutes before quitting. I will leave this running for longer to see if it reaches completion.

I have tried replacing the above line of code as you suggested with fitsignal(V,t,r,@dd_gauss,@bg_hom3d,@ex_4pdeer); and get the following message (below)- however this does produce a plot at the end of it.

In fitsignal/Vmodel (line 672)
  In fitsignal>@(par)Vmodel([],par,i) (line 397)
  In jacobianest (line 105)
  In fitsignal (line 397)
  In untitled (line 19)
Warning: Rank deficient, rank = 0, tol =  0.000000e+00.

On 29 May 2020, at 16:43, Luis Fabregas <notifications@github.com<mailto:notifications@github.com>> wrote:



Ok, I have the feeling that the issue here is that in the algorithm of lmlsqnonlin.m there are some values which are dependent on the precision of the computer. Some of these values are then being fed through a sequence of function handles which end up leading to NaN values. These conditions would explain the erratic behavior of the error and the fact that it is computer-dependent.

I might have traced the error and have made a modification. @ShonaHepworth<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FShonaHepworth&data=02%7C01%7Cs.hepworth%40uea.ac.uk%7C4378f6385f7648d1b4ce08d803e7015f%7Cc65f8795ba3d43518a070865e5d8f090%7C0%7C0%7C637263637978865126&sdata=wj%2FOJSOg7J%2BlPxG9tkhfL352zlNz8KcO7GYkyvz8bt4%3D&reserved=0> I must to ask for your help again. Could you please:

  1.  Download this new lmlsqnonlin.zip<https://eur01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FJeschkeLab%2FDeerLab%2Ffiles%2F4702769%2Flmlsqnonlin.zip&data=02%7C01%7Cs.hepworth%40uea.ac.uk%7C4378f6385f7648d1b4ce08d803e7015f%7Cc65f8795ba3d43518a070865e5d8f090%7C0%7C0%7C637263637978865126&sdata=Th%2BjxS%2F1DoAgBls9I0F67ZoneAAefLA29Md4jBatfjk%3D&reserved=0> and extract again the lmlsqnonlin.m file

  2.  Copy it into the DeerLab\functions\private folder (replacing the old one)

  3.  Run your script with the data that was leading to an error before

  4.  If the error persists try to replace in your script the line

 fitsignal(V,t,r,'P',@bg_hom3d,@ex_4pdeer);


by the line

 fitsignal(V,t,r,@dd_gauss,@bg_hom3d,@ex_4pdeer);


and please let me know it that leads to the same error again. (If not it will point me out to where the error is actually ocurring)

Thank you very much for you help!!

**Comment by ShonaHepworth** [comment]
Hi Luis,

The error no longer occurs. I have double checked this with a couple of other data sets that didn't work before too, so I believe you have found the solution!

Thankyou for all your help

**Comment by luisfabib** [comment]
Yes, I was just writing you back. 

I was running your script again and it took 24min to find a rather bad solution, but without error. Since you could run for over 20min without encountering the error and now you also checked with the other data it seems that the error has been fixed. 

The issue with the long computation times has probably to do with the specific numerics of the problem your script is solving. This is still an issue that I might work on before a release, since it is probably related to too strong convergence criteria of the algorithm. I would also handle the warning you were getting when using the `@dd_gauss` instead of `'P'`, which is a normal problem when the results are not good. 

I really appreciate your help, thank you! 

P.S. A recommendation: do not use ``correctscale()`` anymore. We deprecated this function in `0.9.1` and labeled it as as legacy function, as now all fit functions optimize that internally which leads to better results. I made a small modification to you script which helps getting a "reasonable" solution for the data you sent me in 5s instead of 24min (on my computer). 

````matlab 
clear,clc;

[t,V] = deerload('FileName.DSC'); %File name
axis tight, grid on;
ylabel('(V(t) (AU)'),xlabel('tus');

V = correctphase(V);
t = correctzerotime(V,t);

r = linspace(0,8,200);

tic
 fitsignal(V,t,r,'P',@bg_hom3d,@ex_4pdeer);
toc
````

I will close the issue once I am done with the handling of the warnings and the slow runtime. The complete modifications to `lslsqnonlin.m` will be included in the next patch.

**Comment by luisfabib** [comment]
After further inspection of `lmlsqnonlin.m` I realized that by disabling the Broyden updates of the Jacobian, and instead enforcing full updating of the Jacobian leads to much faster convergence times for the example script above (6s instead of 25min). 

This will be disabled from now on and expect much faster convergence of badly (numerically) conditioned problems. For other cases this should not make a difference.

**Comment by luisfabib** [comment]
To get even more speed out of the algorithm, I improved the dampening of the Levenberg-Marquardt coefficient and added an explicit check if the relative function change to break the iterations.

**Comment by luisfabib** [comment]
All these changes have been released with patch 0.9.2. Issue closed.
